### PR TITLE
Add support for writing NaN/Infinity if SIMDJSON_ENABLE_NAN_INF is enabled

### PR DESCRIPTION
**Pull request by codeinred**
Summary: Updates the string builder and the DOM serialization logic to write out non-finite floats as either `NaN` or `[-]Infinity`

### Description

I added conditional support for NaN/Infinity to the builder and DOM serialization logic when `SIMDJSON_ENABLE_NAN_INF=1`.

I also added tests for this feature, and fixed an edge case related to parsing root Infinity atoms:

```
commit 81ddf93c0c2965e9da734fc15b161b6a216188e8
Author: Alecto Irene Perez <alecto.perez@voladynamics.com>
Date:   Wed May 6 18:21:50 2026 -0400

    Fix bug where 'Inf' as a root atom + spaces of padding parses incorrectly
```

### Relevant Issues

Issues such as #1540, #2540, and #2414, which were interested in parsing NaN/Infinity, may also be interested in having a parser which writes it out.

### Type of change

- [x] Bug fix
- [ ] Optimization
- [x] New feature
- [ ] Refactor / cleanup
- [ ] Documentation / tests
- [ ] Other (please describe):

### How to verify / test

Additional tests were added for this feature under `tests/dom/nan_inf_tests.cpp` and in `tests/builder/builder_string_builder_tests.cpp`.

### Checklist for submitting

- [x] I added/updated tests covering my change (if applicable)
- [x] Code builds locally and passes my check
- [ ] Documentation / README updated if needed
- [x] Commits are atomic and messages are clear
- [x] I linked the related issue (if applicable)

I will have a separate PR with documentation for this change, and for #2696.

### AI Disclosure

Claude assisted with writing/updating the tests to check for this feature.

### CODE REVIEW

File: include/simdjson/generic/atomparsing.h [hunk 1]
```diff
@@ -147,7 +147,9 @@ simdjson_inline bool is_valid_inf_in_string(const uint8_t *src) {
 simdjson_warn_unused
 simdjson_inline bool is_valid_inf_atom(const uint8_t *src, size_t len) {
   if (len > 8) { return is_valid_inf_atom(src); }
-  if (len == 8) { return str8ncmp_case_insensitive(src, "infinity") == 0; }
+  if (len == 8 && str8ncmp_case_insensitive(src, "infinity") == 0) {
```
Comment (lemire) [inline include/simdjson/generic/atomparsing.h:150]: Why this change?

Comment (codeinred) [reply include/simdjson/generic/atomparsing.h:150]: If you have a root atom that is exactly length 8, but contains padding and white space characters, eg `Inf     `, it would result in a tape error instead of correctly matching the shorter Inf atom. It's an edge case that's unlikely to ever come up, but I figured I might as well fix it after I noticed it.

Comment (lemire) [reply include/simdjson/generic/atomparsing.h:150]: I don't understand your answer. But anyhow, I don't think it is important.
